fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/AbstractService.java

```diff
@@ -0,0 +1,290 @@
+     */
+    protected String determineInteractionModel(final List<String> linkTypes,
+                                                   final boolean isRdfContentType, final boolean contentPresent,
+                                                   final boolean isExternalContent) {
+        final String interactionModel = linkTypes == null ? null : linkTypes.stream().filter(INTERACTION_MODELS_FULL::contains).findFirst()
+                .orElse(null);
+
+        // If you define a valid interaction model, we try to use it.
+        if (interactionModel != null) {
+            return interactionModel;
+        }
+        if (isExternalContent || (contentPresent && !isRdfContentType)) {
+            return NON_RDF_SOURCE.toString();
+        } else {
+            return DEFAULT_INTERACTION_MODEL.toString();
+        }
+    }
+
+    /**
+     * Find the closest ancestor using a Fedora ID.
+     * @param fedoraId The fedora ID to check
+     * @return The ancestor ID or null if root
+     */
+    protected String findExistingAncestor(final String fedoraId) {
+        // Remove the ID prefix as it can include a /
+        final String removePrefix = fedoraId.substring(FEDORA_ID_PREFIX.length());
+        if (removePrefix.contains("/")) {
+            return FEDORA_ID_PREFIX + removePrefix.substring(0, removePrefix.lastIndexOf("/"));
+        }
+        return null;
+    }
+
+    /**
+     * Check that we don't try to provide an ACL Link header.
+     * @param links list of the link headers provided.
+     * @throws RequestWithAclLinkHeaderException If we provide an rel="acl" link header.
+     */
+    protected void checkAclLinkHeader(final List<String> links) throws RequestWithAclLinkHeaderException {
+        final var matcher = Pattern.compile("rel=[\"']?acl[\"']?").asPredicate();
+        if (links != null && links.stream().anyMatch(matcher)) {
+            throw new RequestWithAclLinkHeaderException(
+                    "Unable to handle request with the specified LDP-RS as the ACL.");
+        }
+    }
+
+    /**
+     * Check if a path has a segment prefixed with fedora:
+     *
+     * @param externalPath the path.
+     */
+    protected void hasRestrictedPath(final String externalPath) {
+        final String[] pathSegments = externalPath.split("/");
+        if (Arrays.stream(pathSegments).anyMatch(p -> p.startsWith("fedora:"))) {
+            throw new ServerManagedTypeException("Path cannot contain a fedora: prefixed segment.");
+        }
+    }
+
+    /**
+     * Looks through an RdfStream for rdf:types in the LDP namespace or server managed predicates.
+     *
+     * @param model The RDF model.
+     */
+    protected void checkForSmtsLdpTypes(final Model model) {
+        final StmtIterator it = model.listStatements();
+        while (it.hasNext()) {
+            final Statement st = it.next();
+            if ((st.getPredicate().hasURI(RDF_TYPE) && st.getObject().isURIResource() &&
```

Comment:
Sorry, I just noticed this. There is an existing RDF.type constant that you can use, and do an equality comparison here since it is a Property.